ports/esp32/machine_can.c

```diff
@@ -0,0 +1,672 @@
+} filter_mode_t;
+
+typedef struct _machine_can_config_t {
+    can_timing_config_t timing;
+    can_filter_config_t filter;
+    can_general_config_t general;
+    uint16_t baudrate;
+    bool initialized;
+} machine_can_config_t;
+
+typedef struct _machine_can_obj_t {
+    mp_obj_base_t base;
+    machine_can_config_t *config;
+    mp_obj_t rxcallback;
+    byte rx_state;
+    bool extframe : 1;
+    bool loopback : 1;
+    uint16_t num_error_warning; //FIXME: populate this value somewhere
+    uint16_t num_error_passive;
+    uint16_t num_bus_off;
+} machine_can_obj_t;
+
+typedef enum _rx_state_t {
+    RX_STATE_FIFO_EMPTY = 0,
+    RX_STATE_MESSAGE_PENDING,
+    RX_STATE_FIFO_FULL,
+    RX_STATE_FIFO_OVERFLOW,
+} rx_state_t;
+
+// Default baudrate: 500kb
+#define CAN_DEFAULT_PRESCALER (8)
+#define CAN_DEFAULT_SJW (3)
+#define CAN_DEFAULT_BS1 (15)
+#define CAN_DEFAULT_BS2 (4)
+
+// Internal Functions
+mp_obj_t machine_hw_can_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args);
+STATIC mp_obj_t machine_hw_can_init_helper(machine_can_obj_t *self, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args);
+STATIC void machine_hw_can_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind);
+
+// singleton CAN device object
+machine_can_config_t can_config = { .general = CAN_GENERAL_CONFIG_DEFAULT(2, 4, 0),
+                                    .filter = CAN_FILTER_CONFIG_ACCEPT_ALL(),
+                                    .timing = CAN_TIMING_CONFIG_25KBITS(),
+                                    .initialized = false
+                                  };
+
+STATIC machine_can_obj_t machine_can_obj = { {&machine_can_type}, .config = &can_config };
+
+STATIC can_status_info_t _machine_hw_can_get_status() {
+    can_status_info_t status;
+    uint32_t err_code = can_get_status_info(&status);
+    if (err_code != ESP_OK) {
+        mp_raise_OSError(-err_code);
+    }
+    return status;
+}
+
+STATIC void _machine_hw_can_set_filter(machine_can_obj_t *self, uint32_t addr, uint32_t mask, uint8_t bank, bool rtr) {
+    //Check if bank is allowed
+    if ( bank < 0 && bank > ((self->extframe && self->config->filter.single_filter) ? 0 : 1 )) {
+        mp_raise_ValueError("CAN filter parameter error");
+    }
+    uint32_t preserve_mask;
+    if (self->extframe) {
+        addr = (addr & 0x1FFFFFFF) << 3 | (rtr ? 0x04 : 0);
+        mask = (mask & 0x1FFFFFFF) << 3 | 0x03;
+        preserve_mask = 0;
+    } else if (self->config->filter.single_filter) {
+        addr = ( ( (addr & 0x7FF) << 5 ) | (rtr ? 0x10 : 0) );
+        mask = ( (mask & 0x7FF) << 5 );
+        mask |= 0xFFFFF000;
+        preserve_mask = 0;
+    } else {
+        addr = ( ( (addr & 0x7FF) << 5 ) | (rtr ? 0x10 : 0) );
+        mask = ( (mask & 0x7FF) << 5 );
+        preserve_mask = 0xFFFF << ( bank == 0 ? 16 : 0 );
+        if ( (self->config->filter.acceptance_mask & preserve_mask) == ( 0xFFFF << (bank == 0 ? 16 : 0) ) ) {
```

Comment:
why not `(self->config->filter.acceptance_mask & preserve_mask) == preserve_mask` ?